File: web/shared/protocol.ts
// Wire protocol shared by the relay (Deno) and the Cockpit (browser).
// Mirrored in Python at dimos/web/relay_bridge/protocol.py and pinned by the
// golden vectors in ./fixtures/ (tested from both deno test and pytest).
//
// Framing (see web/README.md for the upstream-bug rationale):
// - Control stream frame: u32-LE length | UTF-8 JSON.
// - Datagram: raw UTF-8 JSON, no length prefix.
// - Data frame (one message per stream): u32-LE headerLen | u32-LE payloadLen
//   | header JSON | payload. Receivers count bytes and must never treat
//   stream EOF as a message boundary (Deno 2.6.x delays FIN by up to ~1 s).
//
// Validation policy (mirrored in protocol.py): decoders validate shape
// strictly, and receivers drop invalid or unknown messages -- a peer's bytes
// must never kill a session. Framing-level corruption (absurd length
// prefixes) throws and kills only the affected stream.

export const PROTOCOL_VERSION = 1;

// Reject absurd header lengths before allocating.
export const MAX_HEADER_LEN = 65536;

// Upper bound for a whole data frame; guards receivers against buffering a
// hostile/corrupt payloadLen (also the relay's ingress cap).
export const MAX_DATA_FRAME_BYTES = 64 * 1024 * 1024;

export type Role = "robot" | "viewer";
export type Delivery = "latest" | "reliable";

export interface HelloMsg {
  t: "hello";
  v: number;
  role: Role;
}

export interface WelcomeMsg {
  t: "welcome";
  v: number;
}

export interface PingMsg {
  t: "ping";
  n: number;
  ts: number;
}

export interface PongMsg {
  t: "pong";
  n: number;
  ts: number;
}

export interface ErrorMsg {
  t: "error";
  code: string;
  message: string;
}

// Teleop datagrams (carried from T6 on; declared here so the wire format is
// pinned by fixtures from day one).
export interface TwistMsg {
  t: "twist";
  vx: number;
  wz: number;
  seq: number;
  ts: number;
}

export interface StopMsg {
  t: "stop";
  seq: number;
  ts: number;
}

export type ControlMsg = HelloMsg | WelcomeMsg | PingMsg | PongMsg | ErrorMsg;
export type TeleopMsg = TwistMsg | StopMsg;
export type Msg = ControlMsg | TeleopMsg;

// Data-plane frame header. `delivery` tells the relay how to forward the
// frame without a manifest (T1 only; the T2+ manifest replaces it). `meta`
// carries encoding-specific extras (e.g. {w, h} for images).
export interface FrameHeader {
  ch: string;
  seq: number;
  ts: number;
  delivery: Delivery;
  meta?: Record<string, unknown>;
}

export interface DataFrame {
  header: FrameHeader;
  payload: Uint8Array;
}

const enc = new TextEncoder();
// fatal: reject invalid UTF-8 like Python does, instead of U+FFFD-substituting
// corrupted bytes into channel names or message fields.
const dec = new TextDecoder("utf-8", { fatal: true });

// Runtime field validation, mirror of _MSG_FIELD_KINDS in protocol.py:
// "string" is a JSON string, "number" any JSON number (booleans excluded by
// typeof).
const MSG_FIELDS: Record<string, Record<string, "string" | "number">> = {
  hello: { v: "number", role: "string" },
  welcome: { v: "number" },
  ping: { n: "number", ts: "number" },
  pong: { n: "number", ts: "number" },
  error: { code: "string", message: "string" },
  twist: { vx: "number", wz: "number", seq: "number", ts: "number" },
  stop: { seq: "number", ts: "number" },
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Validated message from parsed JSON; null for unknown or malformed ones. */
export function msgFromUnknown(value: unknown): Msg | null {
  if (!isRecord(value) || typeof value.t !== "string") return null;
  const fields = MSG_FIELDS[value.t];
  if (fields === undefined) return null;
  for (const [name, kind] of Object.entries(fields)) {
    const actual = typeof value[name];
    if (actual !== kind) return null;
  }
  return value as unknown as Msg;
}

/** Validated data-frame header from parsed JSON; null if malformed. */
export function frameHeaderFromUnknown(value: unknown): FrameHeader | null {
  if (
    isRecord(value) &&
    typeof value.ch === "string" &&
    typeof value.seq === "number" &&
    typeof value.ts === "number" &&
    (value.delivery === "latest" || value.delivery === "reliable") &&
    (value.meta === undefined || isRecord(value.meta))
  ) {
    return value as unknown as FrameHeader;
  }
  return null;
}

export function encodeControlFrame(msg: Msg): Uint8Array {
  const body = enc.encode(JSON.stringify(msg));
  const out = new Uint8Array(4 + body.length);
  new DataView(out.buffer).setUint32(0, body.length, true);
  out.set(body, 4);
  return out;
}

/**
 * Incremental parser for a control stream (frames may split across chunks).
 * Malformed or unknown messages are dropped with a console line (the length
 * prefix keeps framing intact); framing errors still throw.
 */
export class ControlFrameReader {
  #buf = new Uint8Array(0);

  push(chunk: Uint8Array): Msg[] {
    const merged = new Uint8Array(this.#buf.length + chunk.length);
    merged.set(this.#buf, 0);
    merged.set(chunk, this.#buf.length);
    this.#buf = merged;
    const msgs: Msg[] = [];
    while (this.#buf.length >= 4) {
      const len = new DataView(this.#buf.buffer, this.#buf.byteOffset).getUint32(0, true);
      if (len === 0 || len > MAX_HEADER_LEN) {
        throw new Error(`invalid control frame length: ${len}`);
      }
      if (this.#buf.length < 4 + len) break;
      const body = this.#buf.subarray(4, 4 + len);
      this.#buf = this.#buf.subarray(4 + len);
      let msg: Msg | null = null;
      try {
        msg = msgFromUnknown(JSON.parse(dec.decode(body)));
      } catch {
        // bad UTF-8 or bad JSON: dropped below
      }
      if (msg === null) console.warn("[protocol] dropping bad control message");
      else msgs.push(msg);
    }
    return msgs;
  }
}

export function encodeDatagram(msg: Msg): Uint8Array {
  return enc.encode(JSON.stringify(msg));
}

/** Returns null for datagrams that are not our JSON messages. */
export function decodeDatagram(data: Uint8Array): Msg | null {
  try {
    return msgFromUnknown(JSON.parse(dec.decode(data)));
  } catch {
    return null;
  }
}

export function encodeDataFrame(header: FrameHeader, payload: Uint8Array): Uint8Array {
  const hdr = enc.encode(JSON.stringify(header));
  const out = new Uint8Array(8 + hdr.length + payload.length);
  const dv = new DataView(out.buffer);
  dv.setUint32(0, hdr.length, true);
  dv.setUint32(4, payload.length, true);
  out.set(hdr, 8);
  out.set(payload, 8 + hdr.length);
  return out;
}

/** Byte lengths of a data frame, or null if fewer than 8 bytes are available. */
export function peekDataFrameLengths(
  buf: Uint8Array,
): { headerLen: number; payloadLen: number; total: number } | null {
  if (buf.length < 8) return null;
  const dv = new DataView(buf.buffer, buf.byteOffset);
  const headerLen = dv.getUint32(0, true);
  const payloadLen = dv.getUint32(4, true);
  if (headerLen > MAX_HEADER_LEN) throw new Error(`data frame header too large: ${headerLen}`);
  const total = 8 + headerLen + payloadLen;
  if (total > MAX_DATA_FRAME_BYTES) throw new Error(`data frame too large: ${total} bytes`);
  return { headerLen, payloadLen, total };
}

export function decodeDataFrame(frame: Uint8Array): DataFrame {
  const lens = peekDataFrameLengths(frame);
  if (lens === null || frame.length < lens.total) {
    throw new Error(`truncated data frame: ${frame.length} bytes`);
  }
  const header = frameHeaderFromUnknown(
    JSON.parse(dec.decode(frame.subarray(8, 8 + lens.headerLen))),
  );
  if (header === null) throw new Error("invalid data frame header");
  return { header, payload: frame.subarray(8 + lens.headerLen, lens.total) };
}

/** First `limit` bytes of `chunks`, copied into one fresh buffer. */
export function concatBytes(chunks: Uint8Array[], limit: number): Uint8Array {
  const out = new Uint8Array(limit);
  let off = 0;
  for (const c of chunks) {
    if (off >= limit) break;
    const take = Math.min(c.byteLength, limit - off);
    out.set(take === c.byteLength ? c : c.subarray(0, take), off);
    off += take;
  }
  return out;
}

/**
 * Incremental reader for a single-message stream. Returns the frame as soon
 * as headerLen + payloadLen bytes have arrived; never waits for EOF. Bytes
 * past the frame are ignored. Chunks are held by reference and copied once
 * at completion (a per-push merge is quadratic at multi-MB frame sizes).
 */
export class DataFrameReader {
  #chunks: Uint8Array[] = [];
  #size = 0;
  #lens: { headerLen: number; payloadLen: number; total: number } | null = null;
  #done = false;

  push(chunk: Uint8Array): DataFrame | null {
    if (this.#done) return null;
    this.#chunks.push(chunk);
    this.#size += chunk.byteLength;
    if (this.#lens === null && this.#size >= 8) {
      this.#lens = peekDataFrameLengths(concatBytes(this.#chunks, 8));
    }
    if (this.#lens === null || this.#size < this.#lens.total) return null;
    this.#done = true;
    const frame = decodeDataFrame(concatBytes(this.#chunks, this.#lens.total));
    this.#chunks = [];
    return frame;
  }
}

File: dimos/web/relay_bridge/protocol.py
# Copyright 2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Wire-protocol mirror of web/shared/protocol.ts.

Pinned by the golden vectors in web/shared/fixtures/ (tested from both pytest
and deno test). Validation runs on pydantic; nothing here needs aioquic or the
rest of the [web] extra.

Framing (see web/README.md for the upstream-bug rationale):
- Control stream frame: u32-LE length | UTF-8 JSON.
- Datagram: raw UTF-8 JSON, no length prefix.
- Data frame (one message per stream): u32-LE headerLen | u32-LE payloadLen |
  header JSON | payload. Receivers count bytes and must never treat stream
  EOF as a message boundary (Deno 2.6.x delays FIN by up to ~1 s).

Validation policy (mirrored in protocol.ts): decoders validate shape strictly,
and receivers drop invalid or unknown messages -- a peer's bytes must never
kill a session. Framing-level corruption (absurd length prefixes) raises
ProtocolError and kills only the affected stream.
"""

from dataclasses import dataclass
import struct
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, ValidationError

from dimos.utils.logging_config import setup_logger

logger = setup_logger()

PROTOCOL_VERSION = 1

# Reject absurd header lengths before allocating (mirrors protocol.ts).
MAX_HEADER_LEN = 65536

# Upper bound for a whole data frame; guards receivers against buffering a
# hostile/corrupt payloadLen (same constant as the relay's ingress cap).
MAX_DATA_FRAME_BYTES = 64 * 1024 * 1024

Role = Literal["robot", "viewer"]
Delivery = Literal["latest", "reliable"]


class ProtocolError(ValueError):
    pass


class _WireModel(BaseModel):
    # strict: no coercion, so a bool or "1" is not a protocol number (mirrors
    # the typeof checks in protocol.ts). allow_inf_nan=False: Python's JSON
    # parser accepts NaN/Infinity where JSON.parse errors, so the mirror must
    # reject them explicitly -- and must refuse to encode them locally.
    model_config = ConfigDict(strict=True, allow_inf_nan=False)


# Number fields are `int | float`, not `float`: JSON has a single number type
# (booleans excluded above), and plain `float` would coerce ints and serialize
# seq=1 as 1.0, breaking byte-exact encoding against the golden fixtures.


class Hello(_WireModel):
    t: Literal["hello"] = "hello"
    v: int | float
    role: Role


class Welcome(_WireModel):
    t: Literal["welcome"] = "welcome"
    v: int | float


class Ping(_WireModel):
    t: Literal["ping"] = "ping"
    n: int | float
    ts: int | float


class Pong(_WireModel):
    t: Literal["pong"] = "pong"
    n: int | float
    ts: int | float


class Error(_WireModel):
    t: Literal["error"] = "error"
    code: str
    message: str


# Teleop datagrams (carried from T6 on; declared so the wire format is pinned
# by fixtures from day one).
class Twist(_WireModel):
    t: Literal["twist"] = "twist"
    vx: int | float
    wz: int | float
    seq: int | float
    ts: int | float


class Stop(_WireModel):
    t: Literal["stop"] = "stop"
    seq: int | float
    ts: int | float


Msg = Hello | Welcome | Ping | Pong | Error | Twist | Stop

# One pydantic-core pass takes raw peer bytes to a validated message: UTF-8
# decoding, JSON parsing, and shape checks together, discriminated on "t".
_MSG_TA: TypeAdapter[Msg] = TypeAdapter(Annotated[Msg, Field(discriminator="t")])


class FrameHeader(_WireModel):
    """Data-plane frame header.

    `delivery` tells the relay how to forward without a manifest (T1 only; the
    T2+ manifest replaces it). `meta` carries encoding-specific extras.
    """

    ch: str
    seq: int | float
    ts: int | float
    delivery: Delivery
    meta: dict[str, Any] | None = None


@dataclass
class DataFrame:
    header: FrameHeader
    payload: bytes


def msg_from_dict(data: dict[str, Any]) -> Msg:
    try:
        return _MSG_TA.validate_python(data)
    except ValidationError as e:
        raise ProtocolError(f"invalid message: {e}") from e


def _msg_from_json(data: bytes) -> Msg:
    try:
        return _MSG_TA.validate_json(data)
    except ValidationError as e:
        raise ProtocolError(f"invalid message: {e}") from e


def encode_control_frame(msg: Msg) -> bytes:
    body = msg.model_dump_json().encode()
    return struct.pack("<I", len(body)) + body


class ControlFrameReader:
    """Incremental parser for a control stream (frames may split across chunks).

    Malformed or unknown messages are dropped with a log line (the length
    prefix keeps framing intact); framing errors still raise ProtocolError.
    """

    def __init__(self) -> None:
        self._buf = bytearray()

    def push(self, chunk: bytes) -> list[Msg]:
        self._buf += chunk
        msgs: list[Msg] = []
        while len(self._buf) >= 4:
            (length,) = struct.unpack_from("<I", self._buf, 0)
            if length == 0 or length > MAX_HEADER_LEN:
                raise ProtocolError(f"invalid control frame length: {length}")
            if len(self._buf) < 4 + length:
                break
            body = bytes(self._buf[4 : 4 + length])
            del self._buf[: 4 + length]
            try:
                msgs.append(_msg_from_json(body))
            except ProtocolError as e:
                logger.warning(f"dropping bad control message: {e}")
        return msgs


def encode_datagram(msg: Msg) -> bytes:
    return msg.model_dump_json().encode()


def decode_datagram(data: bytes) -> Msg | None:
    """Returns None for datagrams that are not our JSON messages."""
    try:
        return _MSG_TA.validate_json(data)
    except ValidationError:
        return None


def encode_data_frame(header: FrameHeader, payload: bytes) -> bytes:
    hdr = header.model_dump_json(exclude_none=True).encode()
    return struct.pack("<II", len(hdr), len(payload)) + hdr + payload


def peek_data_frame_lengths(buf: bytes | bytearray) -> tuple[int, int, int] | None:
    """(headerLen, payloadLen, total) or None if fewer than 8 bytes are available."""
    if len(buf) < 8:
        return None
    header_len, payload_len = struct.unpack_from("<II", buf, 0)
    if header_len > MAX_HEADER_LEN:
        raise ProtocolError(f"data frame header too large: {header_len}")
    total = 8 + header_len + payload_len
    if total > MAX_DATA_FRAME_BYTES:
        raise ProtocolError(f"data frame too large: {total} bytes")
    return header_len, payload_len, total


def decode_data_frame(frame: bytes | bytearray) -> DataFrame:
    lens = peek_data_frame_lengths(frame)
    if lens is None or len(frame) < lens[2]:
        raise ProtocolError(f"truncated data frame: {len(frame)} bytes")
    header_len, _, total = lens
    view = memoryview(frame)
    try:
        header = FrameHeader.model_validate_json(bytes(view[8 : 8 + header_len]))
    except ValidationError as e:
        raise ProtocolError(f"bad data frame header: {e}") from e
    # The payload slice is the only whole-payload copy on the receive path.
    return DataFrame(header=header, payload=bytes(view[8 + header_len : total]))


class DataFrameReader:
    """Incremental reader for a single-message stream.

    Returns the frame as soon as headerLen + payloadLen bytes have arrived;
    never waits for EOF. Bytes past the frame are ignored.
    """

    def __init__(self) -> None:
        self._buf = bytearray()
        self._done = False

    def push(self, chunk: bytes) -> DataFrame | None:
        if self._done:
            return None
        self._buf += chunk
        lens = peek_data_frame_lengths(self._buf)
        if lens is None or len(self._buf) < lens[2]:
            return None
        self._done = True
        frame = decode_data_frame(self._buf)
        self._buf = bytearray()
        return frame
